sourcetool/pkg/auth/implementation.go

```diff
@@ -0,0 +1,191 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+)
+
+//counterfeiter:generate . authenticatorImplementation
+type authenticatorImplementation interface {
+	requestDeviceCode(context.Context) (*DeviceCodeResponse, error)
+	openBrowser(string) error
+	pollForToken(context.Context, string, time.Duration) (string, error)
+	checkTokenStatus(context.Context, string) (string, error)
+	persistToken(token string) error
+	readToken() (string, error)
+}
+
+type defaultImplementation struct{}
+
+// pollForToken polls GitHub periodically until the device flow is done and
+// the token is issued.
+func (di *defaultImplementation) pollForToken(ctx context.Context, deviceCode string, interval time.Duration) (string, error) {
+	for {
+		token, err := di.checkTokenStatus(ctx, deviceCode)
+		if err != nil {
+			if strings.Contains(err.Error(), "authorization_pending") {
+				time.Sleep(interval)
+				continue
+			}
+			if strings.Contains(err.Error(), "slow_down") {
+				interval += 5 * time.Second
+				time.Sleep(interval)
+				continue
+			}
+			return "", err
+		}
+		return token, nil
+	}
+}
+
+// checkTokenStatus checks if the device flow authorization is done and returns
+// the token if ready.
+func (di *defaultImplementation) checkTokenStatus(ctx context.Context, deviceCode string) (string, error) {
+	data := url.Values{}
+	data.Set("client_id", oauthClientID)
+	data.Set("device_code", deviceCode)
+	data.Set("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, tokenURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return "", fmt.Errorf("failed to create token request: %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Accept", "application/json")
+
+	httpClient := &http.Client{Timeout: 30 * time.Second}
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("failed to check token status: %w", err)
+	}
+	defer resp.Body.Close() //nolint:errcheck
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("failed to read token response: %w", err)
+	}
+
+	var tokenResp TokenResponse
+	if err := json.Unmarshal(body, &tokenResp); err != nil {
+		return "", fmt.Errorf("failed to parse token response: %w", err)
+	}
+
+	if tokenResp.Error != "" {
+		return "", fmt.Errorf("%s", tokenResp.Error)
```

Comment:
Add a meaningful error message. As is, `fmt.Errorf` doesn't do anything.

Reply:
Thanks for noticing this, fixed.